### PR TITLE
[WIP / RFC] move generic UID stuff to a seperate class

### PR DESCRIPTION
**Pull request by maggu2810**
Currently the UID is part of the core.thing bundle and states that every UID must start with the binding ID.
To reuse the non-binding, non-thing related UID implementation a new base class for a generic UID has been created.
It is part of the core bundle. So no need to depend on the thing bundle or to start the UID with a binding ID.

### CODE REVIEW

**Comment by triller-telekom** [comment]
I do understand your changes and moving the `UID` class out of the `thing` bundle certainly makes sense. But do you have an example for `UID`s that we use which do not start with the binding ID?

**Comment by maggu2810** [comment]
> But do you have an example for `UID`s that we use which do not start with the binding ID?

Not an Eclipse SmartHome one.
I started using logic same as it is part of the `UID` class (e.g. limit the allowed character of an ID etc.).

If you think there will be no other usage, I can keep it in my bundles, too.
I just thought it would be perhaps also make sense for other consumers.

**Comment by maggu2810** [comment]
Rules are using "simple" strings for its UID, so it seems the rule UID is allowed to contain any character (at least I don't find some limitations in the current code base). I don't know if this has been done by intention or if there should be a RuleUID class (some time), too.

**Comment by sjka** [comment]
I don't have a good use-case in mind currently too. I don't think it's realistic to change the rule ID type, but who knows... 

In any case, I don't have any strong feelings about it. Would be okay for me to do that anyway. @kaikreuzer?

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/UID.java [hunk 1]
```diff
@@ -12,129 +12,83 @@
...
     @Override
     public int hashCode() {
-        final int prime = 31;
-        int result = 1;
-        result = prime * result + Arrays.hashCode(segments);
-        return result;
-    }
-
-    @Override
-    public boolean equals(Object obj) {
```
Comment (htreu) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/UID.java]: out of curiosity: why are you able to remove `equals` here but leave `hashCode` with a call to super?

Comment (htreu) [reply bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/UID.java]: Never mind, I just saw the new `equals` implementation...

Comment (maggu2810) [reply bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/UID.java]: If a bundle uses the `UID` class and import `org.eclipse.smarthome.core.thing` but does not import `org.eclipse.smarthome.core.common` the IDE complains about e.g.:

```text
Access restriction: The method 'GenericUID.hashCode()' is not API (restriction on required project 'org.eclipse.smarthome.core')
```

That has been the reason I added that stubs at the moment.

**Comment by maggu2810** [comment]
For me it does not care if I use an implementation in a separate bundle or in an ESH framework bundle.
If you think that there will be no need for others, no problem to close this PR.

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/UID.java (hunk 1 above)
Comment (htreu) [reply bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/UID.java]: thanks for the explanation.

**Comment by sjka** [comment]
This is really not meant as an objection to this PR. I only can't _think_ of any usages at the moment... 

I mean, we have all sorts on non-thing related entities which have unique identifiers (e.g. voice, transport, sitemap, persistence,...), but they're all using simple strings and there typically is only one instance of them in the system, so no segmented identifiers are required today. 

I still have the feeling that this might become useful one day, but I don't have any proof.

**Comment by kaikreuzer** [comment]
> In any case, I don't have any strong feelings about it. Would be okay for me to do that anyway. @kaikreuzer?

I usually prefer to have stuff refactored only when it is really needed. I agree that the rule engine would be the best candidate for it, but for historic reasons it went for strings, which might be hard to change now.
I am also not too fond of the name "GenericUID". For the super interface, "UID" would imho be the preferred name, but having the same interface in two different ESH packages can also easily lead to confusion.

**Comment by kaikreuzer** [comment]
@maggu2810 What is your suggestion on how to proceed here?

**Comment by maggu2810** [comment]
I agree that GenericUID is not a good name. IMHO the current UID class is a binding specific ID as it requires the first segment to be the binding ID. So perhaps its name should be BindingUID and what I named GenericUID should be UID (as it use case is rather unspecific). Then there could be still the ThingUID as it is and some time in the future (or perhaps never) a RuleUID.

But if you don't think that refactor make sense as changing UID to BindingUID will involve all third party products etc. We can close this.
As long as I am the only one with the "need" to use some basic function in different non-binding related UIDs I can handle it in a custom UID implementation and no need to change the ESH one.

**Comment by kaikreuzer** [comment]
I fully agree that it is unfortunate that the UID class is binding specific and not defined in the core.
But as you say: A refactoring would be a huge breaking change, which I would rather want to avoid, if there isn't really much pain that urges us to do so.
So with your approval, I'd suggest to close this PR.

**Pull request closed without merging**